_posts/2017-04-27-Rust-1.17.md

```diff
@@ -0,0 +1,351 @@
+* [`Result::expect_err`] mirrors `Result::expect`, but with the `Err` case rather than the `Ok` case.
+* [`Cell::swap`] is similar to `std::mem::swap`, but lets you do it with `&Cell` instead of `&mut T`.
+* [`Cell::replace`] is similar to `std::mem::replace`, but lets you do it with `&Cell` instead of `&mut T`.
+* [`Cell::into_inner`] lets you consume the `Cell`, and extract its value.
+* [`Cell::take`] lets you take the value out of a `Cell`, leaving its `Default::default` behind.
+
+[`Arc::from_raw`]: https://doc.rust-lang.org/std/sync/struct.Arc.html#method.from_raw
+[`Arc::into_raw`]: https://doc.rust-lang.org/std/sync/struct.Arc.html#method.into_raw
+[`BTreeMap::range_mut`]: https://doc.rust-lang.org/std/collections/btree_map/struct.BTreeMap.html#method.range_mut
+[`BTreeMap::range`]: https://doc.rust-lang.org/std/collections/btree_map/struct.BTreeMap.html#method.range
+[`Cell::into_inner`]: https://doc.rust-lang.org/std/cell/struct.Cell.html#method.into_inner
+[`Cell::replace`]: https://doc.rust-lang.org/std/cell/struct.Cell.html#method.replace
+[`Cell::swap`]: https://doc.rust-lang.org/std/cell/struct.Cell.html#method.swap
+[`Cell::take`]: https://doc.rust-lang.org/std/cell/struct.Cell.html#method.take
+[`Ordering::then_with`]: https://doc.rust-lang.org/std/cmp/enum.Ordering.html#method.then_with
+[`Ordering::then`]: https://doc.rust-lang.org/std/cmp/enum.Ordering.html#method.then
+[`Rc::from_raw`]: https://doc.rust-lang.org/std/rc/struct.Rc.html#method.from_raw
+[`Rc::into_raw`]: https://doc.rust-lang.org/std/rc/struct.Rc.html#method.into_raw
+[`Result::expect_err`]: https://doc.rust-lang.org/std/result/enum.Result.html#method.expect_err
+[`collections::Bound`]: https://doc.rust-lang.org/std/collections/enum.Bound.html
+[`process::abort`]: https://doc.rust-lang.org/std/process/fn.abort.html
+[`ptr::read_unaligned`]: https://doc.rust-lang.org/std/ptr/fn.read_unaligned.html
+[`ptr::write_unaligned`]: https://doc.rust-lang.org/std/ptr/fn.write_unaligned.html
+
+In other changes, `Cell<T>` used to require that `T: Copy` for many of its methods,
+but [this has been relaxed significantly](https://github.com/rust-lang/rust/pull/39793).
+
+`Box<T>` [now implements](https://github.com/rust-lang/rust/pull/40009) over a dozen new
+conversions with `From`.
+
+`SocketAddr` and `IpAddr` have [some new conversions](https://github.com/rust-lang/rust/pull/39372)
+as well. Previously, you may have written code like this:
+
+```rust
+"127.0.0.1:3000".parse().unwrap()
+```
+
+Now, you can write
+
+```rust
+SocketAddr::from(([127, 0, 0, 1], 3000))
+// or even
+([127, 0, 0, 1], 3000).into())
+``
+
+This removes some uneccesary run-time parsing, and is roughly as readable, depending on
+your preferences.
+
+Backtraces [now have nicer formatting](https://github.com/rust-lang/rust/pull/38165), eliding
+many internal stack frames by default. For example, the full backtrace:
```

Comment:
It looks like maybe this is worded too aggressively https://github.com/rust-lang/rust/pull/40684